engine/execution/ingestion/core.go

```diff
@@ -0,0 +1,444 @@
+	"github.com/onflow/flow-go/engine/execution/state"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/mempool/entity"
+	"github.com/onflow/flow-go/storage"
+	"github.com/onflow/flow-go/utils/logging"
+)
+
+// Core connects the execution components
+// when it receives blocks and collections, it forwards them to the block queue.
+// when the block queue decides to execute blocks, it forwards to the executor for execution
+// when the block queue decides to fetch missing collections, it forwards to the collection fetcher
+// when a block is executed, it notifies the block queue and forwards to execution state to save them.
+type Core struct {
+	unit *engine.Unit // for async block execution
+
+	log zerolog.Logger
+
+	// state machine
+	blockQueue  *block_queue.BlockQueue
+	throttle    Throttle // for throttling blocks to be added to the block queue
+	execState   state.ExecutionState
+	stopControl *stop.StopControl // decide whether to execute a block or not
+
+	// data storage
+	headers     storage.Headers
+	blocks      storage.Blocks
+	collections storage.Collections
+
+	// computation, data fetching, events
+	executor          BlockExecutor
+	collectionFetcher CollectionFetcher
+	eventConsumer     EventConsumer
+}
+
+type Throttle interface {
+	Init(processables chan<- flow.Identifier) error
+	OnBlock(blockID flow.Identifier) error
+	OnBlockExecuted(blockID flow.Identifier, height uint64) error
+}
+
+type BlockExecutor interface {
+	ExecuteBlock(ctx context.Context, block *entity.ExecutableBlock) (*execution.ComputationResult, error)
+}
+
+type EventConsumer interface {
+	BeforeComputationResultSaved(ctx context.Context, result *execution.ComputationResult)
+	OnComputationResultSaved(ctx context.Context, result *execution.ComputationResult) string
+}
+
+func NewCore(
+	logger zerolog.Logger,
+	throttle Throttle,
+	execState state.ExecutionState,
+	stopControl *stop.StopControl,
+	headers storage.Headers,
+	blocks storage.Blocks,
+	collections storage.Collections,
+	executor BlockExecutor,
+	collectionFetcher CollectionFetcher,
+	eventConsumer EventConsumer,
+) *Core {
+	return &Core{
+		log:               logger.With().Str("engine", "ingestion_core").Logger(),
+		unit:              engine.NewUnit(),
+		throttle:          throttle,
+		execState:         execState,
+		blockQueue:        block_queue.NewBlockQueue(logger),
+		stopControl:       stopControl,
+		headers:           headers,
+		blocks:            blocks,
+		collections:       collections,
+		executor:          executor,
+		collectionFetcher: collectionFetcher,
+		eventConsumer:     eventConsumer,
+	}
+}
+
+func (e *Core) Ready() <-chan struct{} {
+	if e.stopControl.IsExecutionStopped() {
+		return e.unit.Ready()
+	}
+
+	e.launchWorkerToConsumeThrottledBlocks()
+
+	return e.unit.Ready()
+}
+
+func (e *Core) Done() <-chan struct{} {
+	return e.unit.Done()
+}
+
+func (e *Core) OnBlock(header *flow.Header, qc *flow.QuorumCertificate) {
```

Comment:
`OnBlock` and `OnCollection` are not supposed to block right? should these just push to a channel?

They could be made using the NewComponentManagerBuilder. This way the fatal error log can instead be a `irrecoverable.SignalerContext.Throw`. See `StopControl.BlockFinalized`

Reply:
>This way the fatal error log can instead be a irrecoverable.SignalerContext.Throw. See StopControl.BlockFinalized

Right, in terms of error handling (using irrecoverable.Throw), I will try in a separate PR along with replacing the old engine.Unit with component manager, since this PR still uses the old engine.Unit.